File: leetCode/993.java
class Solution {
    HashMap<Integer, int[]> nodeDepthAndParentValue = new HashMap<Integer, int[]>();
        
    public boolean isCousins(TreeNode root, int x, int y) {
        
        this.traverse(root, root, 1, x, y);
    
        return (this.nodeDepthAndParentValue.get(x)[0] == this.nodeDepthAndParentValue.get(y)[0] && this.nodeDepthAndParentValue.get(x)[1] != this.nodeDepthAndParentValue.get(y)[1]);
        
    }
    
     public void traverse(TreeNode root, TreeNode parent, int depth, int x, int y) {
            if (root == null) return;
            if (root.val == x || root.val == y) {
                int[] value = new int[2];
                value[0] = depth;
                value[1] = parent.val;
                this.nodeDepthAndParentValue.put(root.val, value);
            };
            this.traverse(root.left, root, depth +1, x, y);
            this.traverse(root.right, root, depth +1, x, y);
        }; 
}

File: leetCode/993.swift
class Solution {
    func isCousins(_ root: TreeNode?, _ x: Int, _ y: Int) -> Bool {
        if root == nil {
            return false
        } else {
            let firstValue = self.traverse(root: root, parent: root!, value: x, depth: 1)
            let secondValue = self.traverse(root: root, parent: root!, value: y, depth: 1)
            if firstValue != nil && secondValue != nil {
                return firstValue![0] == secondValue![0] && firstValue![1] != secondValue![1]
            } else {
                return false
            }
        }
    }
    
    func traverse(root: TreeNode?, parent: TreeNode, value: Int, depth: Int) -> [Int]? {
        if root == nil {
            return nil
        } else if root!.val == value {
            return [depth, parent.val]
        } else {
            if let res = self.traverse(root: root!.left, parent: root!, value: value, depth: depth + 1) {
              return res;  
            } else {
                return self.traverse(root: root!.right, parent: root!, value: value, depth: depth + 1)
            }
            
        }
    }
}